akka-stream/src/main/scala/akka/stream/impl/io/TLSActor.scala

```diff
@@ -368,10 +368,21 @@ import akka.util.ByteString
       log.debug(
         s"wrap: status=${result.getStatus} handshake=$lastHandshakeStatus remaining=${userInBuffer.remaining} out=${transportOutBuffer
           .position()}")
+
     if (lastHandshakeStatus == FINISHED) handshakeFinished()
     runDelegatedTasks()
     result.getStatus match {
       case OK =>
+        // https://github.com/akka/akka/issues/29922
+        // With JDK 11+, it seems to be possible to get the SSLEngine into a state where
+        // result.getStatus == OK && getHandshakeStatus == NEED_WRAP but
+        // it doesn't make any progress any more.
+        //
+        // We guard against this JDK bug by checking for reasonable invariants after the call to engine.wrap
+        require(
+          transportOutBuffer.position() > 0 || lastHandshakeStatus != NEED_WRAP,
```

Comment:
and we are sure that this cannot happen for normal valid requests?

Reply:
Good question. We cannot completely be sure because the exact semantics and invariants of the SSLEngine are not documented accurately enough.

I would say it is a reasonable invariant that is most likely fulfilled with the current implementations right now but the SSLEngine implemented is probably free to change it later to require some rounds of recursion for internal reasons. But I'd say we might want to risk it and fix it later if turns out to be too strict.

Reply:
Agreed